// @progress/kendo-react-taskboard/dist/npm/TaskBoardConfirmDialog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskBoardConfirmDialog = void 0;
var React = require("react");
var PropTypes = require("prop-types");
var kendo_react_dialogs_1 = require("@progress/kendo-react-dialogs");
var kendo_react_buttons_1 = require("@progress/kendo-react-buttons");
/**
 * Represents the TaskBoardConfirmDialog component.
 */
var TaskBoardConfirmDialog = function (props) {
    var onClose = props.onClose, onConfirm = props.onConfirm, dialogMessage = props.dialogMessage, dialogTitle = props.dialogTitle, dialogConfirmButton = props.dialogConfirmButton, dialogCancelButton = props.dialogCancelButton;
    return (React.createElement(kendo_react_dialogs_1.Dialog, { title: dialogTitle, closeIcon: false },
        dialogMessage,
        React.createElement(kendo_react_dialogs_1.DialogActionsBar, { layout: 'end' },
            React.createElement(kendo_react_buttons_1.Button, { themeColor: 'primary', onClick: onConfirm }, dialogConfirmButton),
            React.createElement(kendo_react_buttons_1.Button, { onClick: onClose }, dialogCancelButton))));
};
exports.TaskBoardConfirmDialog = TaskBoardConfirmDialog;
exports.TaskBoardConfirmDialog.propTypes = {
    onClose: PropTypes.func.isRequired,
    onConfirm: PropTypes.func.isRequired,
    dialogMessage: PropTypes.string.isRequired,
    dialogTitle: PropTypes.string.isRequired,
    dialogConfirmButton: PropTypes.string.isRequired,
    dialogCancelButton: PropTypes.string.isRequired
};
exports.TaskBoardConfirmDialog.displayName = 'KendoReactTaskBoardConfirmDialog';
