// @progress/kendo-react-taskboard/dist/es/messages/index.js
var _a;
/**
 * @hidden
 */
export var taskBoardAddColumnButton = 'taskBoard.toolbarAddColumnButton';
/**
 * @hidden
 */
export var taskBoardSearchField = 'taskBoard.toolbarSearchField';
/**
 * @hidden
 */
export var taskBoardEditColumnButton = 'taskBoard.editColumnButton';
/**
 * @hidden
 */
export var taskBoardAddCardButton = 'taskBoard.addCardButton';
/**
 * @hidden
 */
export var taskBoardDeleteColumnButton = 'taskBoard.deleteColumnButton';
/**
 * @hidden
 */
export var taskBoardEditCardButton = 'taskBoard.editCardButton';
/**
 * @hidden
 */
export var taskBoardDeleteCardButton = 'taskBoard.deleteCardButton';
/**
 * @hidden
 */
export var taskBoardDeleteTaskDialogMessage = 'taskBoard.deleteTaskDialogMessage';
/**
 * @hidden
 */
export var taskBoardDeleteTaskDialogTitle = 'taskBoard.deleteTaskDialogTitle';
/**
 * @hidden
 */
export var taskBoardDeleteTaskConfirmButton = 'taskBoard.deleteTaskConfirmButton';
/**
 * @hidden
 */
export var taskBoardDeleteTaskCancelButton = 'taskBoard.deleteTaskCancelButton';
/**
 * @hidden
 */
export var taskBoardAddCardPaneTitle = 'taskBoard.addCardTitle';
/**
 * @hidden
 */
export var taskBoardAddCardTitleLabel = 'taskBoard.addCardTitleLabel';
/**
 * @hidden
 */
export var taskBoardAddCardDescriptionLabel = 'taskBoard.addCardDescriptionLabel';
/**
 * @hidden
 */
export var taskBoardAddCardPriorityLabel = 'taskBoard.addCardPriorityLabel';
/**
 * @hidden
 */
export var taskBoardAddCardCreateButton = 'taskBoard.addCardCreateButton';
/**
 * @hidden
 */
export var taskBoardDelColumnDialogMsg = 'taskBoard.deleteColumnDialogMessage';
/**
 * @hidden
 */
export var taskBoardDelColumnDialogTitle = 'taskBoard.deleteColumnDialogTitle';
/**
 * @hidden
 */
export var taskBoardDelColumnConfirmBtn = 'taskBoard.deleteColumnConfirmButton';
/**
 * @hidden
 */
export var taskBoardDelColumnCancelBtn = 'taskBoard.deleteColumnCancelButton';
/**
 * @hidden
 */
export var taskBoardEditCardPaneTitle = 'taskBoard.editCardPaneTitle';
/**
 * @hidden
 */
export var taskBoardEditCardPaneSaveBtn = 'taskBoard.editCardPaneSaveButton';
/**
 * @hidden
 */
export var taskBoardEditPaneTitleLabel = 'taskBoard.editPaneTitleLabel';
/**
 * @hidden
 */
export var taskBoardEditPaneDescriptionLabel = 'taskBoard.editPaneDescriptionLabel';
/**
 * @hidden
 */
export var taskBoardEditPanePriorityLabel = 'taskBoard.editPanePriorityLabel';
/**
 * @hidden
 */
export var taskBoardEditPaneCancelBtn = 'taskBoard.editPaneCancelButton';
/**
 * @hidden
 */
export var taskBoardPreviewPanePriorityLabel = 'taskBoard.previewPanePriorityLabel';
/**
 * @hidden
 */
export var taskBoardPreviewPaneDeleteBtn = 'taskBoard.previewPaneDeleteButton';
/**
 * @hidden
 */
export var taskBoardPreviewPaneEditBtn = 'taskBoard.previewPaneEditButton';
/**
 * @hidden
 */
export var messages = (_a = {},
    _a[taskBoardAddColumnButton] = 'Add column',
    _a[taskBoardSearchField] = 'Search',
    _a[taskBoardEditColumnButton] = 'Edit column',
    _a[taskBoardAddCardButton] = 'Add card',
    _a[taskBoardDeleteColumnButton] = 'Delete column',
    _a[taskBoardEditCardButton] = 'Edit card',
    _a[taskBoardDeleteCardButton] = 'Delete card',
    _a[taskBoardDeleteTaskDialogMessage] = 'Are you sure you want to delete this card?',
    _a[taskBoardDeleteTaskDialogTitle] = 'Delete Card',
    _a[taskBoardDeleteTaskConfirmButton] = 'Delete',
    _a[taskBoardDeleteTaskCancelButton] = 'Cancel',
    _a[taskBoardDelColumnDialogMsg] = 'Are you sure you want to delete this column?',
    _a[taskBoardDelColumnDialogTitle] = 'Delete column?',
    _a[taskBoardDelColumnConfirmBtn] = 'Delete',
    _a[taskBoardDelColumnCancelBtn] = 'Cancel',
    _a[taskBoardAddCardPaneTitle] = 'Create new card',
    _a[taskBoardAddCardCreateButton] = 'Create',
    _a[taskBoardEditPaneCancelBtn] = 'Cancel',
    _a[taskBoardEditCardPaneTitle] = 'Edit',
    _a[taskBoardEditCardPaneSaveBtn] = 'Save changes',
    _a[taskBoardEditPaneTitleLabel] = 'Title:',
    _a[taskBoardEditPaneDescriptionLabel] = 'Description:',
    _a[taskBoardEditPanePriorityLabel] = 'Priority:',
    _a[taskBoardAddCardTitleLabel] = 'Title',
    _a[taskBoardAddCardDescriptionLabel] = 'Description',
    _a[taskBoardAddCardPriorityLabel] = 'Priority',
    _a[taskBoardPreviewPanePriorityLabel] = 'Priority:',
    _a[taskBoardPreviewPaneDeleteBtn] = 'Delete',
    _a[taskBoardPreviewPaneEditBtn] = 'Edit',
    _a);
