// @progress/kendo-react-taskboard/dist/npm/messages/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.messages = exports.taskBoardPreviewPaneEditBtn = exports.taskBoardPreviewPaneDeleteBtn = exports.taskBoardPreviewPanePriorityLabel = exports.taskBoardEditPaneCancelBtn = exports.taskBoardEditPanePriorityLabel = exports.taskBoardEditPaneDescriptionLabel = exports.taskBoardEditPaneTitleLabel = exports.taskBoardEditCardPaneSaveBtn = exports.taskBoardEditCardPaneTitle = exports.taskBoardDelColumnCancelBtn = exports.taskBoardDelColumnConfirmBtn = exports.taskBoardDelColumnDialogTitle = exports.taskBoardDelColumnDialogMsg = exports.taskBoardAddCardCreateButton = exports.taskBoardAddCardPriorityLabel = exports.taskBoardAddCardDescriptionLabel = exports.taskBoardAddCardTitleLabel = exports.taskBoardAddCardPaneTitle = exports.taskBoardDeleteTaskCancelButton = exports.taskBoardDeleteTaskConfirmButton = exports.taskBoardDeleteTaskDialogTitle = exports.taskBoardDeleteTaskDialogMessage = exports.taskBoardDeleteCardButton = exports.taskBoardEditCardButton = exports.taskBoardDeleteColumnButton = exports.taskBoardAddCardButton = exports.taskBoardEditColumnButton = exports.taskBoardSearchField = exports.taskBoardAddColumnButton = void 0;
/**
 * @hidden
 */
exports.taskBoardAddColumnButton = 'taskBoard.toolbarAddColumnButton';
/**
 * @hidden
 */
exports.taskBoardSearchField = 'taskBoard.toolbarSearchField';
/**
 * @hidden
 */
exports.taskBoardEditColumnButton = 'taskBoard.editColumnButton';
/**
 * @hidden
 */
exports.taskBoardAddCardButton = 'taskBoard.addCardButton';
/**
 * @hidden
 */
exports.taskBoardDeleteColumnButton = 'taskBoard.deleteColumnButton';
/**
 * @hidden
 */
exports.taskBoardEditCardButton = 'taskBoard.editCardButton';
/**
 * @hidden
 */
exports.taskBoardDeleteCardButton = 'taskBoard.deleteCardButton';
/**
 * @hidden
 */
exports.taskBoardDeleteTaskDialogMessage = 'taskBoard.deleteTaskDialogMessage';
/**
 * @hidden
 */
exports.taskBoardDeleteTaskDialogTitle = 'taskBoard.deleteTaskDialogTitle';
/**
 * @hidden
 */
exports.taskBoardDeleteTaskConfirmButton = 'taskBoard.deleteTaskConfirmButton';
/**
 * @hidden
 */
exports.taskBoardDeleteTaskCancelButton = 'taskBoard.deleteTaskCancelButton';
/**
 * @hidden
 */
exports.taskBoardAddCardPaneTitle = 'taskBoard.addCardTitle';
/**
 * @hidden
 */
exports.taskBoardAddCardTitleLabel = 'taskBoard.addCardTitleLabel';
/**
 * @hidden
 */
exports.taskBoardAddCardDescriptionLabel = 'taskBoard.addCardDescriptionLabel';
/**
 * @hidden
 */
exports.taskBoardAddCardPriorityLabel = 'taskBoard.addCardPriorityLabel';
/**
 * @hidden
 */
exports.taskBoardAddCardCreateButton = 'taskBoard.addCardCreateButton';
/**
 * @hidden
 */
exports.taskBoardDelColumnDialogMsg = 'taskBoard.deleteColumnDialogMessage';
/**
 * @hidden
 */
exports.taskBoardDelColumnDialogTitle = 'taskBoard.deleteColumnDialogTitle';
/**
 * @hidden
 */
exports.taskBoardDelColumnConfirmBtn = 'taskBoard.deleteColumnConfirmButton';
/**
 * @hidden
 */
exports.taskBoardDelColumnCancelBtn = 'taskBoard.deleteColumnCancelButton';
/**
 * @hidden
 */
exports.taskBoardEditCardPaneTitle = 'taskBoard.editCardPaneTitle';
/**
 * @hidden
 */
exports.taskBoardEditCardPaneSaveBtn = 'taskBoard.editCardPaneSaveButton';
/**
 * @hidden
 */
exports.taskBoardEditPaneTitleLabel = 'taskBoard.editPaneTitleLabel';
/**
 * @hidden
 */
exports.taskBoardEditPaneDescriptionLabel = 'taskBoard.editPaneDescriptionLabel';
/**
 * @hidden
 */
exports.taskBoardEditPanePriorityLabel = 'taskBoard.editPanePriorityLabel';
/**
 * @hidden
 */
exports.taskBoardEditPaneCancelBtn = 'taskBoard.editPaneCancelButton';
/**
 * @hidden
 */
exports.taskBoardPreviewPanePriorityLabel = 'taskBoard.previewPanePriorityLabel';
/**
 * @hidden
 */
exports.taskBoardPreviewPaneDeleteBtn = 'taskBoard.previewPaneDeleteButton';
/**
 * @hidden
 */
exports.taskBoardPreviewPaneEditBtn = 'taskBoard.previewPaneEditButton';
/**
 * @hidden
 */
exports.messages = (_a = {},
    _a[exports.taskBoardAddColumnButton] = 'Add column',
    _a[exports.taskBoardSearchField] = 'Search',
    _a[exports.taskBoardEditColumnButton] = 'Edit column',
    _a[exports.taskBoardAddCardButton] = 'Add card',
    _a[exports.taskBoardDeleteColumnButton] = 'Delete column',
    _a[exports.taskBoardEditCardButton] = 'Edit card',
    _a[exports.taskBoardDeleteCardButton] = 'Delete card',
    _a[exports.taskBoardDeleteTaskDialogMessage] = 'Are you sure you want to delete this card?',
    _a[exports.taskBoardDeleteTaskDialogTitle] = 'Delete Card',
    _a[exports.taskBoardDeleteTaskConfirmButton] = 'Delete',
    _a[exports.taskBoardDeleteTaskCancelButton] = 'Cancel',
    _a[exports.taskBoardDelColumnDialogMsg] = 'Are you sure you want to delete this column?',
    _a[exports.taskBoardDelColumnDialogTitle] = 'Delete column?',
    _a[exports.taskBoardDelColumnConfirmBtn] = 'Delete',
    _a[exports.taskBoardDelColumnCancelBtn] = 'Cancel',
    _a[exports.taskBoardAddCardPaneTitle] = 'Create new card',
    _a[exports.taskBoardAddCardCreateButton] = 'Create',
    _a[exports.taskBoardEditPaneCancelBtn] = 'Cancel',
    _a[exports.taskBoardEditCardPaneTitle] = 'Edit',
    _a[exports.taskBoardEditCardPaneSaveBtn] = 'Save changes',
    _a[exports.taskBoardEditPaneTitleLabel] = 'Title:',
    _a[exports.taskBoardEditPaneDescriptionLabel] = 'Description:',
    _a[exports.taskBoardEditPanePriorityLabel] = 'Priority:',
    _a[exports.taskBoardAddCardTitleLabel] = 'Title',
    _a[exports.taskBoardAddCardDescriptionLabel] = 'Description',
    _a[exports.taskBoardAddCardPriorityLabel] = 'Priority',
    _a[exports.taskBoardPreviewPanePriorityLabel] = 'Priority:',
    _a[exports.taskBoardPreviewPaneDeleteBtn] = 'Delete',
    _a[exports.taskBoardPreviewPaneEditBtn] = 'Edit',
    _a);
