// @progress/kendo-react-taskboard/dist/es/TaskBoardConfirmDialog.js
import * as React from 'react';
import * as PropTypes from 'prop-types';
import { Dialog, DialogActionsBar } from '@progress/kendo-react-dialogs';
import { Button } from '@progress/kendo-react-buttons';
/**
 * Represents the TaskBoardConfirmDialog component.
 */
export var TaskBoardConfirmDialog = function (props) {
    var onClose = props.onClose, onConfirm = props.onConfirm, dialogMessage = props.dialogMessage, dialogTitle = props.dialogTitle, dialogConfirmButton = props.dialogConfirmButton, dialogCancelButton = props.dialogCancelButton;
    return (React.createElement(Dialog, { title: dialogTitle, closeIcon: false },
        dialogMessage,
        React.createElement(DialogActionsBar, { layout: 'end' },
            React.createElement(Button, { themeColor: 'primary', onClick: onConfirm }, dialogConfirmButton),
            React.createElement(Button, { onClick: onClose }, dialogCancelButton))));
};
TaskBoardConfirmDialog.propTypes = {
    onClose: PropTypes.func.isRequired,
    onConfirm: PropTypes.func.isRequired,
    dialogMessage: PropTypes.string.isRequired,
    dialogTitle: PropTypes.string.isRequired,
    dialogConfirmButton: PropTypes.string.isRequired,
    dialogCancelButton: PropTypes.string.isRequired
};
TaskBoardConfirmDialog.displayName = 'KendoReactTaskBoardConfirmDialog';
